mozci/console/commands/check_backfills.py

```diff
@@ -0,0 +1,173 @@
+        self.pushes = make_push_objects(nb=nb_pushes, branch=branch)
+        nb_pushes = len(self.pushes)
+
+        for index, push in enumerate(self.pushes, start=1):
+            self.line(
+                f"<comment>Processing push {index}/{nb_pushes}: {push.push_uuid}</comment>"
+            )
+            backfill_tasks = []
+
+            try:
+                indexed_tasks = list_indexed_tasks(
+                    f"gecko.v2.{push.branch}.revision.{push.rev}.taskgraph.actions"
+                )
+            except requests.exceptions.HTTPError as e:
+                self.line(
+                    f"<error>Couldn't fetch indexed tasks on push {push.push_uuid}: {e}</error>"
+                )
+                continue
+
+            for indexed_task in indexed_tasks:
+                task_id = indexed_task["taskId"]
+                try:
+                    children_tasks = list_dependent_tasks(task_id)
+                except requests.exceptions.HTTPError as e:
+                    self.line(
+                        f"<error>Couldn't fetch dependent tasks of indexed task {task_id} on push {push.push_uuid}: {e}</error>"
+                    )
+                    continue
+
+                for child_task in children_tasks:
+                    task_action = (
+                        child_task.get("task", {}).get("tags", {}).get("action", "")
+                    )
+                    # We are looking for the Treeherder symbol because Sheriffs are
+                    # only interested in backfill-tasks holding the '-bk' suffix in TH
+                    th_symbol = (
+                        child_task.get("task", {})
+                        .get("extra", {})
+                        .get("treeherder", {})
+                        .get("symbol", "")
+                    )
+                    status = child_task.get("status", {})
+                    if task_action == "backfill-task" and th_symbol.endswith("-bk"):
+                        assert status.get(
+                            "taskId"
+                        ), "Missing taskId attribute in backfill task status"
+                        assert status.get(
+                            "taskGroupId"
+                        ), "Missing taskGroupId attribute in backfill task status"
+                        assert status.get(
+                            "state"
+                        ), "Missing state attribute in backfill task status"
+                        backfill_tasks.append(
+                            BackfillTask(
+                                status["taskId"], status["taskGroupId"], status["state"]
+                            )
+                        )
+                    else:
+                        logger.debug(
+                            f"Skipping non-backfill task {status.get('taskId')}"
+                        )
+
+            def group_key(task):
+                return task.group_id
```

Comment:
Backfill tasks belonging to the same backfill action might be running on different task groups, so if we group by task group we will notify at wrong times.

We should instead group backfills by the treeherder symbol, as all backfill tasks related to a given backfill have the same symbol.

Right @Archaeopteryx?

Reply:
Yes, the notification should only be sent if the tasks in all groups have finished.

Reply:
FYI: With this refactoring, the current task will be indexed under `project.mozci.check-backfill.<environment>.<branch>.<revision>.<treeherder_symbol>` when a notification is sent